File: src/pubsub_py/pubsub_py/pub.py
# rclpy 모듈 및 std_msgs모듈 import
import rclpy  # rclpy는 ros2와 상호작용하는 파이썬 api를 제공하는 모듈
from rclpy.node import Node
from rclpy.qos import QoSProfile

from std_msgs.msg import String


class Pub(Node):  # publisher 노드 클래스 Node클래스 상속
    def __init__(self):  # pub 클래스 생성자
        super().__init__('pub')  # super 함수를 통해 Node 클래스의 생성자에 노드 이름을 매게변수로 작성

        qos_profile = QoSProfile(depth=10)
        # String타입의 메시지를 'topic'이름의 토픽으로 publish함 ,
        # create_publisher 함수를 사용해 publisher 생성, 매개변수로는 메시지 타입, 토픽이름, QoS임
        self.publisher_ = self.create_publisher(String, 'topic', qos_profile)
        timer_period = 0.5  # second 타이머 주기
        # create_timer함수를 통해 타이머 생성, 매개변수는 타이머 주기, 콜백함수
        self.timer = self.create_timer(timer_period, self.timer_callback)
        self.count = 0

    def timer_callback(self):
        msg = String()
        msg.data = 'Hello World: %d' % self.count
        self.publisher_.publish(msg)  # 위의 msg를 publish함수를 사용하여 publish함
        self.get_logger().info('Publishing: "%s"' % msg.data)  # 콘솔 로그 표시
        self.count += 1


def main(args=None):
    rclpy.init(args=args)  # 주어진 args에 대한 ROS통신 초기화
    node = Pub()  # pub 클래스
    try:
        rclpy.spin(node)  # 프로그램이 종료될 때 까지 큐에 요청된 콜백함수 처리
    except KeyboardInterrupt:
        node.get_logger().into('keyboard Interrupt (SIGINT)')
    finally:
        node.destroy_node()  # 노드 종료 시 destroy_node 함수를 통해 publish 노드 소멸
        rclpy.shutdown()  # shutdown 함수를 통해 노드 종료


if __name__ == '__main__':
    main()

File: src/pubsub_cpp/src/pub.cpp
#include <chrono>  // 시간 관련 라이브러리
#include <functional>  // 함수 객체 관련 라이브러리
#include <memory>  // 스마트 포인터 관련 라이브러리
#include <string>  // 문자열 관련 라이브러리

#include "rclcpp/rclcpp.hpp"  // ROS 2 C++ 클라이언트 라이브러리
#include "std_msgs/msg/string.hpp"  // ROS 2 표준 메시지 라이브러리

using namespace std::chrono_literals;  // 시간 리터럴 사용을 위한 네임스페이스

// @brief Pub 클래스는 퍼블리셔 노드를 생성하고 초기화하는 역할을 담당합니다.
class Pub : public rclcpp::Node{ //Node 클래스를 상속받아 Pub 클래스를 정의
    public:
        Pub() : Node("pub"), count(0) //Node 클래스의 생성자를 호출하고 노드 이름을 pub로 설정
        {
            auto qos_profile = rclcpp::QoS(rclcpp::KeepLast(10)); //메시지 큐의 크기를 10개로 설정
            publisher_ = this->create_publisher<std_msgs::msg::String>("topic", qos_profile); //토픽 이름은 topic, 메시지 큐의 크기는 10개
            timer = this->create_wall_timer(500ms, std::bind(&Pub::timer_callback, this)); //500ms마다 timer_callback 함수를 호출
        }

    private:
        void timer_callback() //timer_callback 함수 정의
        {
            auto msg = std_msgs::msg::String(); //std_msgs::msg::String 타입의 메시지 객체 생성
            msg.data = "Hello, world! " + std::to_string(count++); //메시지 객체의 data 멤버에 문자열을 저장
            RCLCPP_INFO(this->get_logger(), "Publishing: '%s'", msg.data.c_str()); //메시지를 출력
            publisher_->publish(msg); //메시지를 퍼블리시
        }
        rclcpp::TimerBase::SharedPtr timer; //타이머 객체
        rclcpp::Publisher<std_msgs::msg::String>::SharedPtr publisher_; //퍼블리셔 객체
        size_t count;
        
};

int main(int argc, char * argv[])
{
    rclcpp::init(argc, argv); //ROS 2 라이브러리 초기화, 매개변수는 main 함수의 매개변수를 그대로 전달
    auto node = std::make_shared<Pub>(); //Pub 클래스의 객체를 생성
    rclcpp::spin(node); //노드 실행
    rclcpp::shutdown(); //노드 종료
    return 0; 
}